src/scheduler/Scheduler.java

```diff
@@ -1,22 +1,133 @@
 package scheduler;
+import java.util.LinkedList;
+import java.util.Queue;
+
 import common.Message;
-import common.Requestable;
-import common.Respondable;
+import common.MessageType;
 
 /**
- * Scheduler will create a thread for each thread
- * and back to listening for more threads.
+ * This class represents a Scheduler. Scheduler will try to schedule 
+ * if there are pending messages.
  * 
  * @author Michael Fan
  *
  */
-public class Scheduler implements Requestable {
-	public void request(Message msg, Respondable response) {
-		System.out.println("Scheduler: Message received from " + msg.getType());
-		System.out.println("Scheduler: Message is: " + msg.getBody());
+public class Scheduler implements Runnable {
+	private Queue<Message> messages;
+	private Queue<Message> elevatorMessages;
+	private Queue<Message> floorMessages;
+	
+	public Scheduler() {
+		messages = new LinkedList<Message>();
+		elevatorMessages = new LinkedList<Message>();
+		floorMessages = new LinkedList<Message>();
+	}
+	
+	/**
+	 * Write a message into the message queue. The client will call
+	 * this method to write a message into the queue.
+	 * 
+	 * @param message the client's message
+	 */
+	public void request(Message message) {
+		System.out.println("Scheduler: Message received from " + message.getType());
+		System.out.println("Scheduler: Message is: " + message.getBody());
 		
-		Thread request = new Thread(new Request(msg, response));
+		synchronized (messages) {
+			messages.add(message);
+		}
+		
+		messages.notifyAll();
+	}
+	
+	/**
+	 * Return messages of the specified type.
+	 * 
+	 * @param messageType the type of the message to return
+	 * 
+	 * @return a list of messages
+	 */
+	public Queue<Message> response(MessageType messageType) {
+		Queue<Message> messages = null;
 		
-		request.start();
+		if(messageType == MessageType.ELEVATOR) {
+			synchronized (elevatorMessages) {
+				while(elevatorMessages.isEmpty()) {
+					try {
+						elevatorMessages.wait();
+					} catch (Exception e) {
+						// TODO: handle exception
+					}
+				}
+				
+				messages = new LinkedList<Message>(elevatorMessages);
+				elevatorMessages.clear();
+				
+				elevatorMessages.notifyAll();
+			}
+		} else if(messageType == MessageType.FLOOR) {
+			synchronized (floorMessages) {
+				while(floorMessages.isEmpty()) {
+					try {
+						floorMessages.wait();
+					} catch (Exception e) {
+						// TODO: handle exception
+					}
+				}
+				
+				messages = new LinkedList<Message>(floorMessages);
+				floorMessages.clear();
+				
+				floorMessages.notifyAll();
+			}
+		}
+		
+		return messages;
+	}
+	
+	/**
+	 * Tries to schedule based on the requests. Will wait if there
+	 * is no requests.
+	 */
+	public void run() {
+		while(!messages.isEmpty()) {
```

Comment:
Can you clarify this statement? When we first initialize Scheduler, wouldn't messages be empty? So this loop would immediately return.